src/DataStreams/RemoteQueryExecutor.cpp

```diff
@@ -406,32 +414,18 @@ void RemoteQueryExecutor::finish(std::unique_ptr<ReadContext> * read_context)
 
     /// Send the request to abort the execution of the request, if not already sent.
     tryCancel("Cancelling query because enough data has been read", read_context);
-
-    /// Get the remaining packets so that there is no out of sync in the connections to the replicas.
-    Packet packet = connections->drain();
-    switch (packet.type)
     {
-        case Protocol::Server::EndOfStream:
-            finished = true;
-            break;
-
-        case Protocol::Server::Log:
-            /// Pass logs from remote server to client
-            if (auto log_queue = CurrentThread::getInternalTextLogsQueue())
-                log_queue->pushBlock(std::move(packet.block));
-            break;
-
-        case Protocol::Server::Exception:
-            got_exception_from_replica = true;
-            packet.exception->rethrow();
-            break;
-
-        default:
-            got_unknown_packet_from_replica = true;
-            throw Exception(ErrorCodes::UNKNOWN_PACKET_FROM_SERVER, "Unknown packet {} from one of the following replicas: {}",
-                toString(packet.type),
-                connections->dumpAddresses());
+        /// Finish might be called in multiple threads. Make sure we release connections in thread-safe way.
+        std::lock_guard guard(connection_draining_mutex);
+        if (auto conn = ConnectionCollector::enqueueConnectionCleanup(pool, std::move(connections)))
```

Comment:
Now I see why we check `if(connections)` inside `ConnectionCollector::enqueueConnectionCleanup`.
Because we could have called `finish` already from another thread, and `connections` can be nullptr.
It is a little bit implicit, but could be ok (maybe add a comment? :)

Reply:
I thought that comment above conveys the message... )